accounting/invoice/models.py

```diff
@@ -20,3 +20,376 @@
+            |   |   |- 12
+            |   |   |   |- invoice_1.pdf
+            |   |   |   |- ...
+            |   |   |   |- invoice_n.pdf
+            |- 2018
+
+        TODO: This should be a synchronous job on a non-web worker.
+        """
+        if not settings.ENABLE_GOOGLE:
+            return
+
+        # Generate a query we'll use repeatedly for finding the proper folder.
+        query = ("mimeType ='application/vnd.google-apps.folder' and "
+                 "'{}' in parents and "
+                 "trashed=false")
+
+        def get_folder_id(root, name):
+            """ Returns the ID of the folder with name `name` in the folder given by ID `root`. """
+            folders = drive.ListFile({'q': query.format(root)}).GetList()
+            for folder in folders:
+                if folder['title'] == name:
+                    return folder['id']
+
+        gauth = GoogleAuth()
+        gauth.ServiceAuth()
+        drive = GoogleDrive(gauth)
+        year_folder = get_folder_id(settings.GOOGLE_DRIVE_ROOT, str(self.billing_start_date.year))
+        invoice_folder = get_folder_id(year_folder, 'invoices-in')
+        month_folder = get_folder_id(invoice_folder, str(self.billing_start_date.month))
+        file = drive.CreateFile({
+            'title': invoice_file.split('/')[-1],
+            'parents': [{'id': month_folder}]
+        })
+        file.SetContentFile(invoice_file)
+        file.Upload()
+
+    def save(self, **kwargs):
+        """
+        Convert the invoice to a PDF and/or upload it to Google Drive after a successful save.
+        """
+        super().save(**kwargs)
+        if self.auto_download_jira_worklogs_on_save:
+            self.fill_line_items_from_jira()
+        if self.auto_pdf_on_save:
+            pdf = self.to_pdf()
+            if self.auto_upload_google_drive_on_save:
+                self.upload_to_google_drive(pdf)
+
+
+class LineItem(models.Model):
+    """
+    A model to hold information related to line items to be billed through an invoice.
+    """
+
+    JIRA_TAG = 'jira_worklog'
+
+    invoice = models.ForeignKey(
+        Invoice, on_delete=models.CASCADE, related_name='line_items',
+        help_text=_("The invoice to which this line item belongs."))
+    line_item_id = models.IntegerField(
+        help_text=_("A non-auto-incrementing ID used for line items with the same fields for when the "
+                    "items may be the same or different. "
+                    "For example: a line item may have the exact same descriptive fields but still be "
+                    "a separate physical item, in which case a UUID or auto-incrementing ID would not suffice "
+                    "to tell the difference."))
+    key = models.CharField(
+        max_length=100,
+        help_text=_("The key identifier for this line item. "
+                    "For example: OC-9999."))
+    name = models.CharField(
+        max_length=255,
+        help_text=_("What the item is."))
```

Comment:
@UmanShahzad How about:

> What the item is. In case of a JIRA worklog, this would be the summary/title of the corresponding ticket.

Reply:
I would like to keep it generic. I know I gave examples like 'OC-9999', but that could be anything's key rather than just JIRA. I will give a generic example, though.